components/raftstore/src/store/worker/pd.rs

```diff
@@ -1084,6 +1170,233 @@ fn send_destroy_peer_message(
     }
 }
 
+const DEFAULT_QPS_THRESHOLD: u32 = 500;
+const DETECT_TIMES: u32 = 10;
+pub const TOP_N: usize = 10;
+const DETECT_INTERVAL: Duration = Duration::from_secs(1);
+const MIN_SAMPLE_NUM: i32 = 100;
+const DEFAULT_SPLIT_SCORE: f64 = 0.25;
+
+pub struct Sample {
+    pub key: Vec<u8>,
+    pub left: i32,
+    pub contained: i32,
+    pub right: i32,
+}
+
+impl Sample {
+    fn new(key: &[u8]) -> Sample {
+        Sample {
+            key: key.to_owned(),
+            left: 0,
+            contained: 0,
+            right: 0,
+        }
+    }
+}
+
+pub struct KeyRange {
+    pub start_key: Vec<u8>,
+    pub end_key: Vec<u8>,
+    pub qps: u64,
+}
+
+impl KeyRange {
+    fn new(start_key: &[u8], end_key: &[u8]) -> KeyRange {
+        KeyRange {
+            start_key: start_key.to_owned(),
+            end_key: end_key.to_owned(),
+            qps: 0,
+        }
+    }
+}
+
+pub struct Recorder {
+    pub samples: Vec<Sample>,
+    pub times: u32,
+    pub count: u64,
+    pub create_time: SystemTime,
+}
+
+impl Recorder {
+    fn new() -> Recorder {
+        Recorder {
+            samples: vec![],
+            times: 0,
+            count: 0,
+            create_time: SystemTime::now(),
+        }
+    }
+
+    fn record(&mut self, key_ranges: &[KeyRange]) {
+        self.times += 1;
+        for key_range in key_ranges.iter() {
+            self.count += 1;
+            if self.samples.len() < 20 {
+                self.samples.push(Sample::new(&key_range.start_key));
+            } else {
+                self.sample(key_range);
+                let i = rand::thread_rng().gen_range(0, self.count) as usize;
+                if i < 20 {
+                    self.samples[i] = Sample::new(&key_range.start_key);
+                }
+            }
+        }
+    }
+
+    fn sample(&mut self, key_range: &KeyRange) {
+        for mut sample in self.samples.iter_mut() {
+            if sample.key.cmp(&key_range.start_key) == Ordering::Less {
+                sample.left += 1;
+            } else if !key_range.end_key.is_empty()
```

Comment:
If it's a point get, is `end_key` empty?
Then `sample.right` always stays still.